testflinger_agent/__init__.py

```diff
@@ -26,6 +32,89 @@
 logger = logging.getLogger(__name__)
 
 
+class ReqBufferTimer(Timer):
+    """Requests buffer flush"""
+
+    def run(self):
+        """Loop timer"""
+        while not self.finished.wait(self.interval):
+            self.function(*self.args, **self.kwargs)
+
+
+class ReqBufferHandler(logging.Handler):
+    """Requests logging handler"""
+
+    def __init__(self, agent, server):
+        super().__init__()
+        if not server.lower().startswith("http"):
+            server = "http://" + server
+        uri = urljoin(server, "/v1/agents/data/")
+        self.url = urljoin(uri, agent)
+        self.qdepth = 100  # messages
+        self.reqbuffer = deque([], maxlen=self.qdepth)
+        self.reqbuff_timer = None
+        self.reqbuff_interval = 10.0  # seconds
+        self._start_reqbuff_timer()
+        # reuse socket
+        self.session = self._requests_retry()
+
+    def _requests_retry(self, retries=3):
+        """Retry api server"""
+        session = requests.Session()
+        retry = Retry(
+            total=retries,
+            read=retries,
+            connect=retries,
+            backoff_factor=0.3,
+            status_forcelist=(500, 502, 503, 504),
+            method_whitelist=False,  # allow post retry
+        )
+        adapter = HTTPAdapter(max_retries=retry)
+        session.mount("http://", adapter)
+        session.mount("https://", adapter)
+        return session
+
+    def _start_reqbuff_timer(self):
+        """Periodically check and send buffer"""
+        self.reqbuff_timer = ReqBufferTimer(self.reqbuff_interval, self.flush)
+        # terminate timer on exit
+        self.reqbuff_timer.daemon = True
+        self.reqbuff_timer.start()
+
+    def emit(self, record):
+        """Write logging events to buffer"""
+        if len(self.reqbuffer) >= self.qdepth:
+            self.reqbuffer.popleft()
+
+        self.reqbuffer.append(record)
+
+    def flush(self):
+        """Flush and post buffer"""
+        try:
+            # atomic queue iteration
+            for record in list(self.reqbuffer):
```

Comment:
needed to extract dequeue when iterating to avoid RuntimeError: deque mutated during iteration